### ISSUE TITLE
The CUDA arena expands to the whole card and nothing in the public API can influence it

### ISSUE THREAD

**Issue by Eric-Louchez** [Shorokoo/Shorokoo#326]
Two things that together make device memory unmanageable on the GPU backend.

**1. The arena expands to the entire card and stays there.** ONNX Runtime's BFC arena extends in
large blocks and does not return them. Training a 49,214,208-parameter decoder-only transformer
(6 layers, width 384, 6 heads of 64, sequence length 1024, vocabulary 50257, fp32, `AdamWOptimizer`)
at batch 8 on an RTX 4090 with 24,563 MiB usable:

```
idle before build: 1,575 MiB of 24,563 MiB
after rig build:   1,579 MiB
step 0: foreground-after 12,877 MiB
step 1: foreground-after 24,563 MiB     <- one expansion of ~11.7 GiB
step 2: foreground-after 24,563 MiB
...
```

It trains happily in that state, but there is no headroom for anything else and no number reported by
the framework that would say how close to the edge a configuration is. At batch 4 the same model
starts at 8,765 MiB and roughly doubles over ~100 steps before going flat, with unchanged live
tensors — arena growth, not the model needing more:

| run | step 0 | later | settled |
|---|---:|---:|---|
| `baseline`, batch 4, 16 steps | 8,765 MiB | 15,963 MiB | still rising at end of run |
| `bov-all` (164.1 M params), batch 4, 305 steps | 9,791 MiB | 17,983 MiB | flat from ~step 100 |
| `baseline`, batch 4, 5,780 steps | — | 15,965 MiB | flat from ~step 200 |

The steady figure is roughly 1.8× the step-0 figure, which is not something a user can predict from
the model.

**2. Nothing in the public API can influence it.** No memory-budget argument, no session-options
surface, no arena-strategy setting. `training.md` is explicit that `ComputeContext` — the only place
such a thing could plausibly live — carries nothing:

> **What the two can usefully differ in: nothing, today.** `ComputeContext` has a single
> parameterless constructor and carries no per-instance settings — no device, no execution provider,
> no thread count, no session options …

The only levers documented anywhere are architectural: `[Module(Checkpoint = true)]`, `queryChunks` on
`Attention.ScaledDotProductAttention` (added for #197), or "shorten the sequence, shrink the batch,
cut heads". All of them change the model or the recipe rather than the allocator.

## What it cost

The recipe called for batch 8 with one configuration shared by every run. The 164.1 M-parameter
variant genuinely does not fit at batch 8 on this card — step 1 fails with

```
Microsoft.ML.OnnxRuntime.OnnxRuntimeException: [ErrorCode:Fail]
onnxruntime::BFCArena::AllocateRawInternal Failed to allocate memory for requested buffer of size 2359296
```

— so every run moved to batch 4 with the step count doubled and the schedule milestones rescaled: a
controlled experiment's training recipe changed for an allocator reason rather than a modelling one,
and batch 4 runs 24% slower per token than batch 8 for the variants that do fit (19,869 vs 24,601
tokens/s).

Before switching, real effort went into restructuring the LM head to remove two avoidable full-size
copies of the `[B·L, 50257]` logits tensor (1.5 GiB per copy at batch 8) by flattening on the small
side of the matmul and folding the soft-cap's `1/15` into the head weight — an exact algebraic
identity. That took the batch-8 step-0 peak from 16,973 MiB to 12,893 MiB and bought nothing, because
it is the arena's *later* expansion that decides whether a configuration fits.

What is missing is any way to ask for a smaller arena, state a budget, or learn how close a
configuration is to the edge. Reading the figure at all needs a hand-written `cudaMemGetInfo`
P/Invoke — #198.

Related: #198 (no device-memory visibility), and the error-message half of this on a Job-Object-capped
process.

**No migration work.** Shorokoo has no installed base to protect. Add the settings surface wherever it
belongs and change `ComputeContext` or the backend factory signature outright — no compatibility shim
and no version bump.
